torch/distributed/checkpoint/_state_dict_utils.py

```diff
@@ -86,7 +107,11 @@ def _gather_state_dict(
             )
             value = value.to_local()
         elif isinstance(value, dict):
-            value = _gather_state_dict(value, pg, device)
+            value = _gather_state_dict(value, pg=pg, device=device)
+
+        if isinstance(value, torch.Tensor) and cpu_offload:
+            value = value.to(cpu_device)
 
-        new_state_dict[key] = value
+        if not cpu_offload or len(ranks_only) == 0 or dist.get_rank(pg) in ranks_only:
```

Comment:
Is `dist.get_rank(pg) in ranks_only` along not sufficient?

Reply:
ranks_only can be empty (the default value). But it seems that we should not restrict this function to be cpu_offload only.

Reply:
I see. Makes sense!